build/org.eclipse.pde.build/src/org/eclipse/pde/internal/build/builder/SourceGenerator.java

```diff
@@ -463,20 +472,29 @@ private void replaceXMLAttribute(String location, String tag, String attr, Strin
 				continue;
 			}
 
-			int startVersionId = Utils.scan(buffer, startAttributeWord + 1, "\""); //$NON-NLS-1$
-			int endVersionId = Utils.scan(buffer, startVersionId + 1, "\""); //$NON-NLS-1$
+			int startVersionId = buffer.indexOf("\"", startAttributeWord + 1); //$NON-NLS-1$
+			int endVersionId = buffer.indexOf("\"", startVersionId + 1); //$NON-NLS-1$
 			buffer.replace(startVersionId + 1, endVersionId, newValue);
 			attrFound = true;
 		}
 		if (attrFound) {
 			try {
-				Files.writeString(featureFile.toPath(), buffer.toString());
+				Files.writeString(featureFile, buffer.toString(), fileCharset);
 			} catch (IOException e) {
 				//ignore
 			}
 		}
 	}
 
+	private static Charset getFileEncoding(Path file) throws IOException, XMLStreamException {
+		try (InputStream is = Files.newInputStream(file)) {
+			XMLInputFactory factory = XMLInputFactory.newFactory();
+			factory.setProperty(XMLInputFactory.SUPPORT_DTD, false);
```

Comment:
@jukzi did you recall if PDE features need doctype resolution at all?

Reply:
> On the other hand, since this parser is only used to read the encoding, the parser could not come into the situation to attempt to resolve them at all.

Just checked it and not supporting DTD at all is not a problem even if present, as expected.